### ISSUE TITLE
The weight is killing innovations in `JS` frameworks

### ISSUE THREAD

**Issue by ClassicOldSong** [krausest/js-framework-benchmark#1487]
In short, the weight put `JavaScript` frameworks in an unfair competition.

If this benchmark is called "web-framework-benchmark", I would have no objection on this weight, but... it's called "js-framework-benchmark" so...

The newly added weight encourage frameworks change into "compile time optimizations" that work only in browsers - frameworks who use `innerHTML` and `cloneNode` to gain speed in node creation, but only in browsers, while originally JS frameworks can have a much wider availability in multiple other scenarios such as iOS and Android app development(e.g. with [NativeScript](https://nativescript.org/) with [DOMiNATIVE](https://github.com/SudoMaker/dominative)), and even embedded GUI development(what I'm working on right now), but these environments are very unlikely to provide a built-in XML parser and even if, won't benefit much from doing so. While `document.createElement` can be provided and translated very well in multiple GUI implementations, using the same API also provides a even ground for each frameworks to compete about their "JS" part - `innerHTML` and `cloneNode` are provided by browsers, not JS itself.

Also, in my opinion, node creation speed isn't that important that needs a ~60% weight, but content updating speed affects user experience in a bigger aspect.

ref: #1341 #1233

**Comment by ClassicOldSong** [comment]
I must point the fact out that, JavaScript is not a BROWSER ONLY language, and it can have much wider usages other than browsers and NodeJS. Since JavaScript frameworks have the most versatile experience and accumulation in the GUI domain, we can make them being useful to other fields other than browsers.

But this needs us to think in an open mind. People who didn't realize that JS can do more than browsers, we need to let them know.

**Comment by leeoniya** [comment]
> If this benchmark is called "web-framework-benchmark", I would have no objection on this weight, but... it's called "js-framework-benchmark" so...

yes, effectively this is a "web-framework-benchmark". or if you want to be even more pedantic, it's a "dom-framework-benchmark", since Three.js can also be considered a "web" framework, as can any gui library compiled to WASM + WebGL.

nevertheless, "js framework" has a de-facto meaning of "dom framework", much like "Coke" refers to a drink that no longer contains cocaine. most people recognize this repo for what it is, i think. im not sure if getting hung up on semantics is a worthwhile endeavor.

**Comment by ClassicOldSong** [comment]
Actually DOM framework still isn't accurate - innerHTML isn't strictly a DOM thing

Anyway this weight still encourages frameworks to "optimize" only for browsers and eventually will break their ability to run elsewhere, narrowing their actual usability.

I want to raise awareness of the fact that JS frameworks appear here can have greater potentials, and I don't want the weight to kill them.

**Comment by ClassicOldSong** [comment]
There's already an example that broke: Svelte 4 runs flawlessly in NativeScript and my embedded environment, but Svelte 5 made the innerHTML optimized compiler the only option, rendering it completely unusable in these scenarios.

Solid can be hacked to still use createElement and it's now functioning well in NativeScript, they even did a livestream about this days ago.

Please don't narrow possibilities down.

**Comment by ClassicOldSong** [comment]
You want to keep the name"js-framework-benchmark" is fine, but could we make the other potential about those frameworks visible and noticeable?

**Comment by leeoniya** [comment]
> innerHTML isn't strictly a DOM thing

it quite literally is: https://w3c.github.io/DOM-Parsing/#dom-innerhtml-innerhtml, and has nothing to do with ECMAScript (JS). what you probably _mean_ is that some JS frameworks adopted this well-known DOM getter/setter name for their own concept of "string of markup/html". if a framework's strategy of pre-compilation is not compatible with a non-standard DOM environment, then this sounds like something you should discuss with either the framework authors, or the authors of the environment that is DOMish-but-not-really.

> but Svelte 5 made the innerHTML optimized compiler the only option, rendering it completely unusable in these scenarios.

and if that is faster in the standard DOM environment, then it was probably the correct choice for them, given that's likely the overwhelming majority of Svelte's (Vercel's) target audience.

> but could we make the other potential about those frameworks visible and noticeable?

React has the ability to target non-dom environments (via React Native), and it is plenty visible/competitive here, right?

**Comment by ClassicOldSong** [comment]
I mean, over 90% frameworks here can work in non-browser environments, they just didn't realize that.

DOM wasn't built for HTML in the beginning, what you're referring to is HTML DOM.

RN handles non-dom environment itself, but what I'm talking about is making most frameworks have this ability without creating wheels again and again themselves. This is totally different than RN's approach.

We could just implement a minimal DOM in any new GUI environment and boom we have lots of available frameworks to run already, but if we need to talk to every framework author about writing a specific implementation again? That's not ideal.

**Comment by ClassicOldSong** [comment]
What they could clearly realize is, their score were affected even more with creation speed and they'll tend to use the more browser specific APIs, without the knowledge of they can have more potentials.

This is pretty bad since it's clearly harmful to diversity.

**Comment by leeoniya** [comment]
> I mean, over 90% frameworks here can work in non-browser environments,

i don't really understand what you're trying to say here. with a proper environment shim everything can work everywhere. Wine/Proton exists and all the software built for Windows "just never realized" that it can also run on Linux?...because of Wine?

all of these frameworks can obviously run in non-dom environments as long as those environments have the features they rely on from the DOM. and that is a pretty huge statement to make because this benchmarks measures the cost of doing proper css and html compliant layout, styling, event model, text layout. any shimmed environment that does not provide all of this cannot be compared (in this benchmark) to one that does.

**Comment by ClassicOldSong** [comment]
Idk if you have worked with any embedded environments, but their resources are very limited and mostly will not benefit from having a XML parser built with their runtime.

I've clearly stated this in the beginning.

Shims are fine, until they're not. Like Wine more or less still affects performance, that's why we still need native Linux executables.

**Comment by ClassicOldSong** [comment]
Further more, CSS has nothing to do with these frameworks. They're interfacing DOM only, while browsers handle the CSS property bindings from their DOM.

Providing a minimal enough shim to run these frameworks is not huge and I've already done it on a device that has only 8MB of RAM, with XIP it could be even 2MB. CSS isn't needed since it's not the only styling method in the world.

**Comment by leeoniya** [comment]
have you asked the Svelte authors (or authors of any fast frameworks here) if they would consider reducing their DOM performance and featureset to maintain compatibility with an environment that has a reduced subset of the DOM and no CSS layout? for frameworks that target the browser as their first-class/priority, i'm guessing the answer will universally be "no".

> Further more, CSS has nothing to do with these frameworks. 

but it has something to do with _this benchmark_, because it measures the time required to do proper CSS parsing and layout of the rendered content. if your framework or environment does not render with the same rules as the DOM/CSS, then it is not going to be measured against the rest that do.

> This is pretty bad since it's clearly harmful to diversity.

and Formula 1 cars dont race in NASCAR races. all participants in the benchmark must play by the same rules. in this benchmark, they all must render using the same set of technologies: DOM, CSS, HTML, JS, WASM. they cannot render on Canvas or use WebGL.

**Comment by ClassicOldSong** [comment]
> have you asked the Svelte authors (or authors of any fast frameworks here) if they would consider reducing their DOM performance and featureset to maintain compatibility with an environment that has a reduced subset of the DOM and no CSS layout?

Yes I have and the answer is "yes" by considering adding back the createElement way, though time isn't guaranteed.

What I'm arguing about is don't loose potentials, while providing compile options that do both is even better than having only one choice.

**Comment by ClassicOldSong** [comment]
You still didn't get the biggest issue here, which is people don't know they have more potentials and they will lose them when the benchmark doesn't tell them and encourage them to optimize for only one direction.

The problem here is hiding choices.

**Comment by localvoid** [comment]
> providing compile options that do both is even better than having only one choice.

Why should I choose a subset of an awful DOM API as a universal retained mode UI API?

`innerHTML`/`cloneNode` optimization isn't the only thing that is used to optimize specifically for a browser environment. For example, in the ivi library, Right-To-Left update algorithm or children list diffing that tries to find minimal DOM ops won't be useful outside of a browser environment.

**Comment by ClassicOldSong** [comment]
> Why should I choose a subset of an awful DOM API as a universal retained mode UI API?

Certainly you can choose not to, but you'll lose the ability to run on more platforms.

And, I think you haven't read carefully enough: we're talking about running outside of browsers that can't provide ultimately the same APIs due to multiple reasons, and your quote just said it's even better if you can choose when you need those fancy optimizations, and your code can still work when these optimizations are not available.

**Comment by ClassicOldSong** [comment]
> Right-To-Left update algorithm or children list diffing that tries to find minimal DOM ops won't be useful outside of a browser environment.

Actually they're still usable since it's using only DOM level one APIs, that is available in ANY DOM implementation.

Go check `undom-ng`, which is the actual DOM implementation powering DOMiNATIVE and you'll understand what I mean.

**Comment by localvoid** [comment]
> Actually they're still usable

Usable doesn't mean that they will be useful. If I am going to target different platform, I am going to use different architecture/algorithms/etc and I am sure as hell I don't want to target an awful DOM API when I am not constrained by a browser environment, especially if I am going to optimize for a resource constrained embedded environment.

**Comment by ClassicOldSong** [comment]
I wrote my own frameworks, and I have a new framework that does DOM diffing also(but with my own algorithm), working flawlessly in my DOM impl and in browser, and Solid/Vue/React all ran smoothly in my DOM implementation. DOM itself isn't a browser specific thing but HTML DOM is.
When DOM points to generic XML documents, innerHTML isn't appropriate anymore. And as it's a fact that it's a MODEL, it can actually run without any XML like backend.

I believe ivi might run well on NativeScript writing native iOS and Android apps right now if it's not using the innerHTML trick.

It's up to you frameworks authors to determine if you want to embrace more platforms, or limiting your framework to browser only.

**Comment by ClassicOldSong** [comment]
> I am going to use different architecture/algorithms/etc

Sure thing, it's your choice to build wheels again and again for your own interests, but I am trying to let people taking benefits from their existing work without having to do duplicated labor, and we can maximize the experience and accumulations frontend frameworks have already gained by reusing the existing code and libraries.

And just lacking innerHTML means "awful" is really entertaining to be heard.

**Comment by localvoid** [comment]
> And just lacking innerHTML means "awful" is really entertaining to be heard.

I've never said that "lacking innerHTML means awful" or anything like that.

**Comment by ClassicOldSong** [comment]
> I've never said that "lacking innerHTML means awful" or anything like that.

Then could you clarify what "awful" are you referring to?

**Comment by ClassicOldSong** [comment]
Still the name is not so accurate and the drag race points frameworks to the narrow end - people take this benchmark as gold standard for a framework is good or bad nowadays...

**Comment by ClassicOldSong** [comment]
> To my knoweledge the [reality](https://survey.stackoverflow.co/2023/#web-frameworks-and-technologies) is far away from that.

That's true to framework *users*, but not *authors*. This benchmark means more to framework authors than actual end developers. Since end users don't care how frameworks are implemented, they might have no idea it's using innerHTML or not, but whether their knowledge about a framework can be migrated to somewhere else matters. That's why this benchmark may mislead framework authors to a narrow end matters.

**Comment by ClassicOldSong** [comment]
> A developer is always an user of some piece of technology, as JS in this very case, and occasionally may be also an author of a JS framework

I think there're some survivorship bias happing here: there're plenty of developers that won't even try understanding the technology at all, they just do developing for a living, as they stay "noob" forever, they still can solve problems most non-developer people have. And as all participants in this thread all contributed to this benchmark, I'd say we already stepped out of junior developers and might see the world different. But we're not talking about this problem here.

> Here, I see your issue is reduced to "using innerHTML or not", which is an issue only in a half of this benchmark, and sadly, does not resonate at all with the main title.

Actually this is my main concern with this issue, the "content updating speed" one happens to be magnified by the newly added weight so I mentioned it. Maybe I should open another issue for that one.

**Comment by krausest** [comment]
Cleanup - I'm closing this issue.
